Implement framework for migrating system indices

AthenaEryma opened the pull request:
This PR adds a framework for migrating system indices as necessary prior
to Elasticsearch upgrades. This framework uses REST APIs added in
another commit:
- GET _migration/system_features

This API, which gets the status of "features" (plugins which own system
indices) with regards to whether they need to be upgraded or not. As of
this PR, this API also reports errors encountered while migrating system
indices alongside the index that was being processed when this occurred.

As an example of this error reporting:

```json
{
    "feature_name": "logstash_management",
    "minimum_index_version": "8.0.0",
    "upgrade_status": "ERROR",
    "indices": [
        {
            "index": ".logstash",
            "version": "8.0.0",
            "failure_cause": {
                "error": {
                    "root_cause": [
                        {
                            "type": "runtime_exception",
                            "reason": "whoopsie",
                            "stack_trace": "<omitted for brevity>"
                        }
                    ],
                    "type": "runtime_exception",
                    "reason": "whoopsie",
                    "stack_trace": "<omitted for brevity>"
                }
            }
        }
    ]
}
```

- POST _migration/system_features

This API starts the migration process. The API for this has no changes,
but when called, any system indices which need to be migrated will be
migrated, with status information stored in the cluster state for later
use by the GET _migration/system_features API.

AthenaEryma committed:
Implement framework for migrating system indices

This PR adds a framework for migrating system indices as necessary prior
to Elasticsearch upgrades. This framework uses REST APIs added in
another commit:
- GET _migration/system_features

This API, which gets the status of "features" (plugins which own system
indices) with regards to whether they need to be upgraded or not. As of
this PR, this API also reports errors encountered while migrating system
indices alongside the index that was being processed when this occurred.

As an example of this error reporting:

```json
{
    "feature_name": "logstash_management",
    "minimum_index_version": "8.0.0",
    "upgrade_status": "ERROR",
    "indices": [
        {
            "index": ".logstash",
            "version": "8.0.0",
            "failure_cause": {
                "error": {
                    "root_cause": [
                        {
                            "type": "runtime_exception",
                            "reason": "whoopsie",
                            "stack_trace": "<omitted for brevity>"
                        }
                    ],
                    "type": "runtime_exception",
                    "reason": "whoopsie",
                    "stack_trace": "<omitted for brevity>"
                }
            }
        }
    ]
}
```

- POST _migration/system_features

This API starts the migration process. The API for this has no changes,
but when called, any system indices which need to be migrated will be
migrated, with status information stored in the cluster state for later
use by the GET _migration/system_features API.

AthenaEryma committed:
Remove extra `onResponse`

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

AthenaEryma committed:
Address Todos/cleanup

AthenaEryma committed:
Return the correct list when starting migration

AthenaEryma committed:
Tweak docs

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Fix compilation after merge

williamrandolph reviewed (changes requested):
I've read through all the code so far and I like it. I have a few comments and requests, and I will be happy to review again once we have some good tests. But overall: very impressive!

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Fix test failures that expected fake responses

AthenaEryma committed:
Extract constant for upgrade version per review

AthenaEryma committed:
Javadoc + cleanup per review

AthenaEryma committed:
Rename results classes to clarify their relationship

SystemIndexMigrationResult -> FeatureMigrationResults
FeatureMigrationStatus -> SingleFeatureMigrationResult

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Tests + always run on master node

AthenaEryma committed:
Fix index name & doc count type

AthenaEryma committed:
Javadoc

AthenaEryma committed:
Plumb in callback metadata + test it

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Line length

AthenaEryma committed:
Verify that the pre/post migration hooks are called

AthenaEryma committed:
Set the origin before calling reindex

AthenaEryma committed:
Check results in upgrade callback + fix bug where results were updated too early

AthenaEryma committed:
Add multi-feature integration test

AthenaEryma committed:
Assert both feature names

AthenaEryma committed:
Cleanup

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Fix unit tests for new logic

williamrandolph reviewed (changes requested):
This is looking really, really good. I have some nitpicks and testing requests, but nothing substantial. However, I'm going to mark it as "request changes" anyway so I can get a notification when commits from end-to-end testing make it here.

One thing I want to look at is making sure that, at the end of the process, all the original aliases are pointing at the migrated index.

Great work!

AthenaEryma committed:
Javadoc + missed renames per review

AthenaEryma committed:
Copy aliases + alias assertions

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Add necessary system privileges

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

AthenaEryma committed:
Logging tweaks

AthenaEryma committed:
Fix test for real version math

AthenaEryma committed:
Adjust test for new system permissions

AthenaEryma committed:
Merge branch 'master' into si/upgrade-framework

williamrandolph reviewed (approved):
The code since my last review is very straightforward, and testing a build of the branch worked like a charm.

This is looking really great. Thank you so much for all the good work here.
